File: tests/tilepack_test/tilepack.py
#!/usr/bin/env python3
"""
Tilepack HW → Radio File Builder
--------------------------------
Loads an image from disk, resizes to VGA, tiles it, compresses each tile
as JPEG, and builds a single binary "radio file" that the SAT can
downlink to GS.

Radio file format (image_radio_file.bin):

  record[0] | record[1] | ... | record[N-1]

Each record has fixed size RECORD_BYTES and layout:

  [page_id (u16)][tile_idx (u16)][frag_idx (u8)][frag_cnt (u8)]
  [jpeg_fragment...][0x00 padding...] → RECORD_BYTES bytes total

RECORD_BYTES MUST MATCH the _FILE_PKTSIZE constant on the SATELLITE_RADIO
side (currently 240).
"""

import io
import os
import math
import csv
import struct
import sys
from dataclasses import dataclass
from typing import List, Tuple
from PIL import Image

# Settings
INPUT_IMAGE   = "/home/argus/RidgeRun_test.jpg"  # Path to image on Jetson
OUTPUT_DIR    = "tilepack"                       # Output directory (created if needed)

PAGE_ID       = 1                                # Image/page ID
TARGET_WIDTH  = 640                              # Resize width (VGA)
TARGET_HEIGHT = 480                              # Resize height (VGA)

TILE_W        = 64                               # Tile width   (pixels)
TILE_H        = 32                               # Tile height  (pixels)
JPEG_QUALITY  = 10                               # JPEG quality (1–95)
RECORD_BYTES  = 240                              # MUST MATCH _FILE_PKTSIZE on the SAT mainboard

# ============================================================

# Per-record header (6 bytes)
@dataclass
class PacketHeader:
    page_id: int
    tile_idx: int
    frag_idx: int
    frag_cnt: int

    def to_bytes(self) -> bytes:
        # >HHBB = big-endian: u16, u16, u8, u8
        return struct.pack(">HHBB", self.page_id, self.tile_idx, self.frag_idx, self.frag_cnt)

    @staticmethod
    def size_bytes() -> int:
        return 6


@dataclass
class Packet:
    header: PacketHeader
    payload: bytes

# Helpers
def image_to_tiles(img: Image.Image, tile_w: int, tile_h: int) -> Tuple[List[Image.Image], int, int]:
    """
    Split image into tiles of size (tile_w x tile_h).
    Pads with black if the image is not an exact multiple.
    Returns (tiles_list, tiles_x, tiles_y).
    """
    w, h = img.size
    tiles_x = math.ceil(w / tile_w)
    tiles_y = math.ceil(h / tile_h)

    # Pad if needed
    if (w % tile_w) != 0 or (h % tile_h) != 0:
        pad_w = tiles_x * tile_w
        pad_h = tiles_y * tile_h
        canvas = Image.new("RGB", (pad_w, pad_h), (0, 0, 0))
        canvas.paste(img, (0, 0))
        img = canvas

    tiles: List[Image.Image] = []
    for ty in range(tiles_y):
        for tx in range(tiles_x):
            box = (tx * tile_w, ty * tile_h, (tx + 1) * tile_w, (ty + 1) * tile_h)
            tiles.append(img.crop(box))

    return tiles, tiles_x, tiles_y


def compress_tile_jpeg(tile: Image.Image, jpeg_quality: int) -> bytes:
    """
    Compress a single tile as JPEG and return raw bytes.
    """
    bio = io.BytesIO()
    tile.save(bio, format="JPEG", quality=jpeg_quality, optimize=True)
    return bio.getvalue()


def packetize_tile(page_id: int, tile_idx: int, tile_bytes: bytes) -> List[Packet]:
    """
    Fragment compressed tile bytes into a list of Packets, where each Packet
    will later become one fixed-size record in the radio file.

    payload size per Packet = RECORD_BYTES - header_size (6 bytes).
    """
    hdr_sz = PacketHeader.size_bytes()
    max_payload = max(1, RECORD_BYTES - hdr_sz)

    # Split tile_bytes into chunks that fit into one record's payload space
    frags = [tile_bytes[i : i + max_payload] for i in range(0, len(tile_bytes), max_payload)]
    frag_cnt = len(frags)

    packets: List[Packet] = []
    for i, frag in enumerate(frags):
        header = PacketHeader(page_id=page_id, tile_idx=tile_idx, frag_idx=i, frag_cnt=frag_cnt)
        packets.append(Packet(header=header, payload=frag))

    return packets


# ---------------------------------------------------------------------------------------------
def main():
    # 0) Check input
    if not os.path.exists(INPUT_IMAGE):
        print(f"[ERROR] Input image not found: {INPUT_IMAGE}", file=sys.stderr)
        sys.exit(1)

    # 1) Prepare output dirs
    os.makedirs(OUTPUT_DIR, exist_ok=True)
    tiles_dir = os.path.join(OUTPUT_DIR, "tiles_jpeg")
    os.makedirs(tiles_dir, exist_ok=True)

    # 2) Load and resize
    img = Image.open(INPUT_IMAGE).convert("RGB").resize((TARGET_WIDTH, TARGET_HEIGHT))
    src_png = os.path.join(OUTPUT_DIR, "input_resized.png")
    img.save(src_png)

    # 3) Tile the image
    tiles, tiles_x, tiles_y = image_to_tiles(img, TILE_W, TILE_H)

    # 4) JPEG-compress each tile
    comp_tiles: List[bytes] = []
    for idx, t in enumerate(tiles):
        b = compress_tile_jpeg(t, JPEG_QUALITY)
        comp_tiles.append(b)

        # Save tile JPEG for sanity checks
        with open(os.path.join(tiles_dir, f"tile_{idx:05d}.jpg"), "wb") as f:
            f.write(b)

    # 5) Packetize all tiles into Packet objects (NOT yet fixed-size records)
    packets: List[Packet] = []
    for idx, tb in enumerate(comp_tiles):
        packets.extend(packetize_tile(PAGE_ID, idx, tb))

    # 6) Write image-wide metadata (useful but not required by comms)
    meta_path = os.path.join(OUTPUT_DIR, "image_meta.bin")
    meta_blob = struct.pack(
        ">HHHHHHHB",
        PAGE_ID,
        tiles_x, tiles_y,
        TILE_W, TILE_H,
        TARGET_WIDTH, TARGET_HEIGHT,
        max(0, min(255, JPEG_QUALITY)),
    )
    with open(meta_path, "wb") as mf:
        mf.write(meta_blob)

    # 7) Build and write the RADIO FILE with fixed-size records
    radio_file_path = os.path.join(OUTPUT_DIR, "image_radio_file.bin")
    hdr_sz = PacketHeader.size_bytes()

    # Manifest is optional but nice for debugging
    manifest_csv_path = os.path.join(OUTPUT_DIR, "packets_manifest.csv")

    with open(radio_file_path, "wb") as rf, open(manifest_csv_path, "w", newline="") as mf:
        writer = csv.writer(mf)
        writer.writerow(
            [
                "record_idx",
                "record_size_bytes",
                "header_size_bytes",
                "payload_size_bytes",
                "page_id",
                "tile_idx",
                "frag_idx",
                "frag_cnt",
            ]
        )

        for i, p in enumerate(packets):
            header = p.header.to_bytes()
            payload = p.payload

            # Build the record: header + payload
            record = bytearray(header + payload)

            if len(record) > RECORD_BYTES:
                raise RuntimeError(
                    f"Record {i} exceeds RECORD_BYTES={RECORD_BYTES} (got {len(record)} bytes). "
                    "Increase RECORD_BYTES or reduce JPEG_QUALITY / tile size."
                )

            # Pad with zeros up to RECORD_BYTES
            pad_len = RECORD_BYTES - len(record)
            record += b"\x00" * pad_len

            # Write this fixed-size record to the radio file
            rf.write(record)

            # Manifest row for debugging
            writer.writerow(
                [
                    i,
                    RECORD_BYTES,
                    hdr_sz,
                    len(payload),
                    p.header.page_id,
                    p.header.tile_idx,
                    p.header.frag_idx,
                    p.header.frag_cnt,
                ]
            )

    # 8) Print summary
    comp_total = sum(len(c) for c in comp_tiles)
    max_payload = max(1, RECORD_BYTES - hdr_sz)

    print("=== TILEPACK HW → RADIO FILE BUILDER ===")
    print(f"Input:            {INPUT_IMAGE}  resized={TARGET_WIDTH}x{TARGET_HEIGHT}")
    print(f"Tiling:           {tiles_x} x {tiles_y}  (tile {TILE_W}x{TILE_H}), tiles={len(tiles)}")
    print(f"JPEG quality:     {JPEG_QUALITY}")
    print(f"Compressed total: {comp_total/1024:.1f} kB")
    print(f"Header bytes:     {hdr_sz}")
    print(f"Record bytes:     {RECORD_BYTES} (max payload per record ≈ {max_payload})")
    print(f"Total records:    {len(packets)}")
    print(f"Saved resized:    {src_png}")
    print(f"Saved tiles:      {tiles_dir}/tile_00000.jpg ..")
    print(f"Saved meta:       {meta_path}")
    print(f"Saved radio file: {radio_file_path}")
    print(f"Manifest:         {manifest_csv_path}")


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print(f"[ERROR] {e}", file=sys.stderr)
        sys.exit(1)

File: tests/tilepack_test/tilepack.cpp
#include "tilepack.hpp"
#include <fstream>
#include <algorithm>
#include <map>
#include <cstring>
#include <iostream>
#include <cmath>

#include <opencv2/opencv.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

namespace tilepack {

// ============================================================
// PacketHeader Implementation
// ============================================================

std::vector<uint8_t> PacketHeader::to_bytes() const {
    std::vector<uint8_t> bytes(PACKET_HEADER_SIZE);
    
    bytes[0] = (payload_size_bytes >> 8) & 0xFF;
    bytes[1] = payload_size_bytes & 0xFF;
    bytes[2] = (page_id >> 8) & 0xFF;
    bytes[3] = page_id & 0xFF;
    bytes[4] = (tile_idx >> 8) & 0xFF;
    bytes[5] = tile_idx & 0xFF;
    bytes[6] = frag_idx;
    
    return bytes;
}

PacketHeader PacketHeader::from_bytes(const uint8_t* data, size_t len) {
    if (len < PACKET_HEADER_SIZE) {
        throw std::runtime_error("Insufficient data for packet header");
    }
    
    PacketHeader header;
    header.payload_size_bytes = (static_cast<uint16_t>(data[0]) << 8) | data[1];
    header.page_id = (static_cast<uint16_t>(data[2]) << 8) | data[3];
    header.tile_idx = (static_cast<uint16_t>(data[4]) << 8) | data[5];
    header.frag_idx = data[6];
    
    return header;
}

// ============================================================
// ImageMetadata Implementation
// ============================================================

std::vector<uint8_t> ImageMetadata::to_bytes() const {
    std::vector<uint8_t> bytes(15);
    
    // Big-endian encoding: >HHHHHHHB
    bytes[0] = (page_id >> 8) & 0xFF;
    bytes[1] = page_id & 0xFF;
    bytes[2] = (tiles_x >> 8) & 0xFF;
    bytes[3] = tiles_x & 0xFF;
    bytes[4] = (tiles_y >> 8) & 0xFF;
    bytes[5] = tiles_y & 0xFF;
    bytes[6] = (tile_w >> 8) & 0xFF;
    bytes[7] = tile_w & 0xFF;
    bytes[8] = (tile_h >> 8) & 0xFF;
    bytes[9] = tile_h & 0xFF;
    bytes[10] = (target_width >> 8) & 0xFF;
    bytes[11] = target_width & 0xFF;
    bytes[12] = (target_height >> 8) & 0xFF;
    bytes[13] = target_height & 0xFF;
    bytes[14] = jpeg_quality;
    
    return bytes;
}

ImageMetadata ImageMetadata::from_bytes(const uint8_t* data, size_t len) {
    if (len < 15) {
        throw std::runtime_error("Insufficient data for metadata");
    }
    
    ImageMetadata meta;
    meta.page_id = (static_cast<uint16_t>(data[0]) << 8) | data[1];
    meta.tiles_x = (static_cast<uint16_t>(data[2]) << 8) | data[3];
    meta.tiles_y = (static_cast<uint16_t>(data[4]) << 8) | data[5];
    meta.tile_w = (static_cast<uint16_t>(data[6]) << 8) | data[7];
    meta.tile_h = (static_cast<uint16_t>(data[8]) << 8) | data[9];
    meta.target_width = (static_cast<uint16_t>(data[10]) << 8) | data[11];
    meta.target_height = (static_cast<uint16_t>(data[12]) << 8) | data[13];
    meta.jpeg_quality = data[14];
    
    return meta;
}

// ============================================================
// TilepackEncoder Implementation
// ============================================================

TilepackEncoder::TilepackEncoder(uint16_t page_id,
                                 uint16_t target_width,
                                 uint16_t target_height,
                                 uint16_t tile_w,
                                 uint16_t tile_h,
                                 int jpeg_quality)
    : page_id_(page_id),
      target_width_(target_width),
      target_height_(target_height),
      tile_w_(tile_w),
      tile_h_(tile_h),
      jpeg_quality_(jpeg_quality) {
    
    // Initialize metadata
    metadata_.page_id = page_id;
    metadata_.tile_w = tile_w;
    metadata_.tile_h = tile_h;
    metadata_.target_width = target_width;
    metadata_.target_height = target_height;
    metadata_.jpeg_quality = static_cast<uint8_t>(std::max(0, std::min(255, jpeg_quality)));
}

TilepackEncoder::~TilepackEncoder() {
}

bool TilepackEncoder::load_image(const std::string& image_path) {
    // Load image using OpenCV
    cv::Mat img = cv::imread(image_path, cv::IMREAD_COLOR);
    if (img.empty()) {
        std::cerr << "Failed to load image: " << image_path << std::endl;
        return false;
    }
    
    std::cout << "Loaded image: " << img.cols << "x" << img.rows << std::endl;
    
    // Resize to target dimensions
    cv::Mat resized;
    cv::resize(img, resized, cv::Size(target_width_, target_height_), 0, 0, cv::INTER_LINEAR);
    
    std::cout << "Resized to: " << resized.cols << "x" << resized.rows << std::endl;
    
    // Tile the image
    if (!tile_image(resized)) {
        return false;
    }
    
    // Packetize all tiles
    packetize_tiles();
    
    std::cout << "Generated " << tiles_.size() << " tiles" << std::endl;
    std::cout << "Generated " << packets_.size() << " packets" << std::endl;
    
    return true;
}

bool TilepackEncoder::tile_image(const cv::Mat& image) {
    tiles_.clear();
    
    int width = image.cols;
    int height = image.rows;
    
    // Calculate number of tiles
    int tiles_x = static_cast<int>(std::ceil(static_cast<double>(width) / tile_w_));
    int tiles_y = static_cast<int>(std::ceil(static_cast<double>(height) / tile_h_));
    
    metadata_.tiles_x = static_cast<uint16_t>(tiles_x);
    metadata_.tiles_y = static_cast<uint16_t>(tiles_y);
    
    // Pad image if necessary
    cv::Mat padded_img;
    int padded_width = tiles_x * tile_w_;
    int padded_height = tiles_y * tile_h_;
    
    if (width != padded_width || height != padded_height) {
        padded_img = cv::Mat::zeros(padded_height, padded_width, image.type());
        image.copyTo(padded_img(cv::Rect(0, 0, width, height)));
    } else {
        padded_img = image.clone();
    }
    
    // Extract tiles
    uint16_t tile_idx = 0;
    for (int ty = 0; ty < tiles_y; ty++) {
        for (int tx = 0; tx < tiles_x; tx++) {
            // Extract tile region
            cv::Rect roi(tx * tile_w_, ty * tile_h_, tile_w_, tile_h_);
            cv::Mat tile = padded_img(roi);
            
            // Compress tile as JPEG
            std::vector<uint8_t> jpeg_data;
            if (!compress_tile_jpeg(tile, jpeg_data)) {
                std::cerr << "Failed to compress tile " << tile_idx << std::endl;
                return false;
            }
            
            Tile t;
            t.tile_idx = tile_idx;
            t.jpeg_data = jpeg_data;
            tiles_.push_back(t);
            
            tile_idx++;
        }
    }
    
    return true;
}

bool TilepackEncoder::compress_tile_jpeg(const cv::Mat& tile, std::vector<uint8_t>& jpeg_out) {
    // JPEG compression parameters
    std::vector<int> params;
    params.push_back(cv::IMWRITE_JPEG_QUALITY);
    params.push_back(jpeg_quality_);
    params.push_back(cv::IMWRITE_JPEG_OPTIMIZE);
    params.push_back(1);
    
    // Encode as JPEG
    if (!cv::imencode(".jpg", tile, jpeg_out, params)) {
        std::cerr << "Failed to encode tile as JPEG" << std::endl;
        return false;
    }
    
    return true;
}

void TilepackEncoder::packetize_tiles() {
    packets_.clear();
    
    for (const auto& tile : tiles_) {
        auto tile_packets = packetize_tile(tile.tile_idx, tile.jpeg_data);
        packets_.insert(packets_.end(), tile_packets.begin(), tile_packets.end());
    }
}

std::vector<Packet> TilepackEncoder::packetize_tile(uint16_t tile_idx,
                                                     const std::vector<uint8_t>& tile_bytes) {
    std::vector<Packet> packets;
    
    // Split tile_bytes into fragments that fit within MAX_PAYLOAD_PER_PACKET
    size_t offset = 0;
    uint8_t frag_idx = 0;
    
    while (offset < tile_bytes.size()) {
        size_t remaining = tile_bytes.size() - offset;
        size_t payload_size = std::min(remaining, MAX_PAYLOAD_PER_PACKET);
        
        Packet pkt;
        pkt.header.payload_size_bytes = static_cast<uint16_t>(payload_size);
        pkt.header.page_id = page_id_;
        pkt.header.tile_idx = tile_idx;
        pkt.header.frag_idx = frag_idx;
        
        pkt.payload.assign(tile_bytes.begin() + offset,
                          tile_bytes.begin() + offset + payload_size);
        
        packets.push_back(pkt);
        
        offset += payload_size;
        frag_idx++;
    }
    
    return packets;
}

bool TilepackEncoder::write_radio_file(const std::string& output_path) const {
    std::ofstream file(output_path, std::ios::binary);
    if (!file.is_open()) {
        std::cerr << "Failed to open output file: " << output_path << std::endl;
        return false;
    }
    
    for (const auto& pkt : packets_) {
        // Write header
        auto header_bytes = pkt.header.to_bytes();
        file.write(reinterpret_cast<const char*>(header_bytes.data()), header_bytes.size());
        
        // Write payload
        file.write(reinterpret_cast<const char*>(pkt.payload.data()), pkt.payload.size());
        
        // Verify packet size
        if (pkt.total_size() > MAX_PACKET_SIZE) {
            std::cerr << "Warning: Packet exceeds MAX_PACKET_SIZE" << std::endl;
        }
    }
    
    file.close();
    return true;
}

bool TilepackEncoder::write_metadata(const std::string& output_path) const {
    std::ofstream file(output_path, std::ios::binary);
    if (!file.is_open()) {
        std::cerr << "Failed to open metadata file: " << output_path << std::endl;
        return false;
    }
    
    auto meta_bytes = metadata_.to_bytes();
    file.write(reinterpret_cast<const char*>(meta_bytes.data()), meta_bytes.size());
    
    file.close();
    return true;
}

size_t TilepackEncoder::get_total_bytes() const {
    size_t total = 0;
    for (const auto& pkt : packets_) {
        total += pkt.total_size();
    }
    return total;
}

size_t TilepackEncoder::get_compressed_size() const {
    size_t total = 0;
    for (const auto& tile : tiles_) {
        total += tile.jpeg_data.size();
    }
    return total;
}

double TilepackEncoder::get_avg_packet_size() const {
    if (packets_.empty()) return 0.0;
    return static_cast<double>(get_total_bytes()) / packets_.size();
}

} // namespace tilepack
